src/callbacks_stage/subcell_bounds_check_2d.jl

```diff
@@ -10,26 +10,28 @@
                               time, iter, output_directory, save_errors)
     (; local_minmax, positivity) = solver.volume_integral.limiter
     (; variable_bounds) = limiter.cache.subcell_limiter_coefficients
-    (; idp_bounds_delta) = limiter.cache
+    (; idp_bounds_delta_local, idp_bounds_delta_global) = limiter.cache
 
     if local_minmax
         for v in limiter.local_minmax_variables_cons
             v_string = string(v)
             key_min = Symbol(v_string, "_min")
             key_max = Symbol(v_string, "_max")
-            deviation_min = idp_bounds_delta[key_min]
-            deviation_max = idp_bounds_delta[key_max]
-            for element in eachelement(solver, cache), j in eachnode(solver),
-                i in eachnode(solver)
-
-                var = u[v, i, j, element]
-                deviation_min[1] = max(deviation_min[1],
-                                       variable_bounds[key_min][i, j, element] - var)
-                deviation_max[1] = max(deviation_max[1],
-                                       var - variable_bounds[key_max][i, j, element])
+            deviation_min_threaded = idp_bounds_delta_local[key_min]
+            deviation_max_threaded = idp_bounds_delta_local[key_max]
+            @threaded for element in eachelement(solver, cache)
+                deviation_min = deviation_min_threaded[Threads.threadid()]
+                deviation_max = deviation_max_threaded[Threads.threadid()]
+                for j in eachnode(solver), i in eachnode(solver)
+                    var = u[v, i, j, element]
+                    deviation_min = max(deviation_min,
+                                        variable_bounds[key_min][i, j, element] - var)
+                    deviation_max = max(deviation_max,
+                                        var - variable_bounds[key_max][i, j, element])
+                end
+                deviation_min_threaded[Threads.threadid()] = deviation_min
+                deviation_max_threaded[Threads.threadid()] = deviation_max
             end
```

Comment:
I noticed a possible problem with this current parallel bounds check routine. When I'm running elixirs with subcell limiting and the subcell bounds checking (it seems to be a bigger issue for small (=only 1 limiters enabled), the maximum-finding part of the bounds check (as marked for the local limiting) is slower with multiple threads and doesn't scale well in general.
```
blast nonperiodic (local limiting of rho, level 6 = 4096 Elements)
            Roci:    My machine:
1 Thread:   470 μs,  344 μs
2 Threads:  490 μs,  210 μs
4 Threads:  475 μs,  166 μs
6 Threads:  450 μs,  200 μs
8 Threads:  580 μs,  188 μs
10 Threads: 300 μs,  116 μs
12 Threads: 340 μs
14 Threads: 295 μs
16 Threads: 290 μs
20 Threads: 220 μs
```

Then, I activated the specific Entropy limiter.
```
blast nonperiodic (local + spec entropy, level 6 = 4096 Elements)
	     Roci:      My machine:
1 Thread:    1.141ms,   896μs
2 Threads:   1.009ms,   568μs
4 Threads:     870μs,   353μs
6 Threads:     824μs,   546μs
8 Threads:   1.115ms,   539μs
10 Threads:    690μs,   447μs
12 Threads:    614μs
20 Threads:    395μs
```
It has the same problems, but slightly less impactful.

I also testes another elixir, the astro jet setup. It uses the same amount of elements and also the local density and spec Entropy limiting.
```
astro jet (local + spec Entropy, level 6 = 4096 Elements)
	    Roci:    My machine:
1 Thread:    2.113ms,   1.342ms
2 Threads:   1.563ms,    769μs
4 Threads:   1.075ms,    430μs
6 Threads:    742μs,     788μs
8 Threads:   1.012ms,    641μs
10 Threads:   754μs,     556μs
12 Threads:   618μs
14 Threads:   542μs
20 Threads:   413μs
```
We again see that the runs with 6 or 8 threads aren't scaling well. But overall, multithreading speeds up the bounds check.
The key difference between the two simulations that the calculation of the entropy `s = entropy_spec(get_node_vars(u, equations, solver, i, j, element), equations)` takes much longer for the astro jet simulation. Probably since the calculation of the entropy takes about twice the time due to  more complex solution vectors over the whole domain.

In summary, I think it's a bit weird that 8 threads (on Roci) and 6 threads (on my machine) work so bad in comparison. But I would suggest to still merge this PR because the bounds check still scale in general, and the problem mostly occurs for small simulations where the doesn't take much time anyway.

Reply:
Are these the CPU times of running `check_bounds`?... If yes, I don't understand why it should get faster or slower when the entropy limiter is activated. As far as I can see, this routine checks the bounds of only the positivity and minmax limiter.

Reply:
Oh sorry. Yes, this is the cpu time of the bounds check. I benchmarked only this routine.
However, since I wanted to test more complex examples I implemented the parallel bounds check on my branch `subcell-limiting`, which includes the bounds check for the spec Entropy limiting. The routine can be found [here](https://github.com/bennibolm/Trixi.jl/blob/0091a0aa5da1451e9869b0f6b85fda009f3f44d1/src/callbacks_stage/subcell_bounds_check_2d.jl#L8).
Sorry for the confusion.

Reply:
@amrueda said that False Sharing could be the problem. As described [here](https://discourse.julialang.org/t/parallel-reductions/30180).
But we don't really have an idea to fix that. Do you have an idea and/or suggestion? @sloede @ranocha

Reply:
You may try `@turbo thread=true`

Reply:
False sharing would explain why your original version didn't have these problems. There you stored a vector for each thread, so the data was scattered across memory. Now, each thread operates on the same contiguous array.

You could try storing a vector of size one for each thread and hope that these small vectors don't lie too close together in memory (or did you try that already? I think I suggested that earlier).

Reply:
> You may try `@turbo thread=true`

Thanks @ranocha. I tried this out.
First with
```
@turbo thread=true for element in eachelement(solver, cache)
    deviation_min = deviation_min_threaded[Threads.threadid()]
    for j in eachnode(solver), i in eachnode(solver)
        var = u[v, i, j, element]
        deviation_min = max(deviation_min, variable_bounds[key_min][i, j, element] - var)
    end
    deviation_min_threaded[Threads.threadid()] = deviation_min
end
```
and got an error message 
```
ERROR: LoadError: UndefVarError: `deviation_min` not defined
```

It seems like the use inside the maximum of a newly defined variable destroys something here. So, I tested it like that:
```
@turbo thread=true for element in eachelement(solver, cache)
    for j in eachnode(solver), i in eachnode(solver)
        var = u[v, i, j, element]
        deviation_min_threaded[Threads.threadid()] = max(deviation_min_threaded[Threads.threadid()], variable_bounds[key_min][i, j, element] - var)
    end
end
```
But, now I get
```
TypeError: in typeassert, expected Int64, got a value of type Nothing
```

Of course, this works without the `@turbo`. I can't really find anything about that error online. Do you have an idea @ranocha ?

Reply:
I can't tell you anything without full stack traces. But you can also try the approach suggested by @efaulhaber above

Reply:
I updated the error messages above.
Yes, thanks. I'm currently testing it using a Vector.

Reply:
You need to update the other stuff as well. See https://juliasimd.github.io/LoopVectorization.jl/stable/examples/multithreading/#Multithreading how to write a multi-threaded reduction with LoopVectorization.jl

Reply:
I'm not really sure what you mean and what my error is.
I want to thread the loop over all elements, but just splitting the memory split over the number of threads. Is that possible with `@turbo` and if yes, how?
Looking at your link, I never really see that `threadid()` is used.

Reply:
That's my point. They set up this stuff for you.

Reply:
I tested @efaulhaber's suggestions [from here](https://github.com/trixi-framework/Trixi.jl/pull/1736#discussion_r1425084965) and used a vector of length 1 instead of just a single number inside the threaded memory. The results were horrible. To make sure false sharing isn't the problem, I then used a vector of length 100 (but still just used the first entry of it). Now, the code is scaling remarkably better (although the time for only a few threads are worse):
```
blast nonperiodic: level 6 = 4096 Elements
		Roci: Normal / Vector(1) / Vector(100)	My machine: Normal / Vector(1) / Vector(100)
1 Thread:   	466 μs  594 μs  598 μs,   		285 μs  378 μs  378 μs
2 Threads:  	503 μs  1.1 ms  322 μs,   		170 μs  495 μs  201 μs
4 Threads:  	470 μs  3.5 ms  164 μs,   		142 μs  475 μs  105 μs
6 Threads:  	321 μs  1.7 ms  110 μs,   		141 μs  811 μs  121 μs
8 Threads:  	610 μs  1.9 ms   84 μs,   		151 μs  518 μs   90 μs
10 Threads: 	401 μs  1.4 ms   69 μs,   		124 μs  473 μs   80 μs
12 Threads: 	347 μs  1.1 ms   58 μs,
16 Threads: 	300 μs  967 μs   50 μs,
```
I'm not really sure what to do with these results. Randomly finding the prefect length of the vector doesn't seem to be a good option :sweat_smile:

Reply:
Aha! So it seems that the problem is indeed false sharing... I would expect that the perfect length of the vector is 8 as the size of a cache line is 64 bytes. Would it be reasonable to allocate the array using something like:
```julia
using CpuId
size_of_array = Int(ceil(cachelinesize()/sizeof(RealT)))
```
@ranocha @sloede?

I find it a bit disappointing that one has to do these things manually in Julia... Also that there are no in-built threading options like the OpenMP schedules or reduction operations.

For the moment, anyway, I think that we should use the `Vector(size_of_array)`version. Even though the serial version gets a bit slower doing that, I think that we should prioritize the threaded performance.

Reply:
On Friday and this morning, I tried out @sloede's suggestion to expand `idp_bounds_delta_local[key]` like
```
idp_bounds_delta_local[key] = [zero(real(basis)) for _ in 1:8*Threads.nthreads()]
```
and then just call every 8th entry like
```
deviation = deviation_threaded[8*Threads.threadid()]
```

The results are great
```
blast nonperiodic: level 6 = 4096 Elements
Roci:            Normal / Vector(1) / Vector(100) / 8*..
1 Thread:   	466 μs  594 μs  598 μs  475 μs,   		
2 Threads:  	503 μs  1.1 ms  322 μs  240 μs,   		
4 Threads:  	470 μs  3.5 ms  164 μs  121 μs,   		
6 Threads:  	321 μs  1.7 ms  110 μs   81 μs,   		
8 Threads:  	610 μs  1.9 ms   84 μs   62 μs,  
10 Threads: 	401 μs  1.4 ms   69 μs   50 μs,  
12 Threads: 	347 μs  1.1 ms   58 μs   41 μs,
16 Threads: 	300 μs  967 μs   50 μs   32 μs,
```
and
```
astro jet (local + spec Entropy, level 6 = 4096 Elements)
Roci:           Normal   /   8*..		
1 Thread:   2.095 ms,  2.149 ms,	
2 Threads:  1.509 ms,  1.123 ms,	
4 Threads:  1.096 ms,    553 μs, 	 
6 Threads:    878 μs,    382 μs, 	 
8 Threads:    949 μs,    281 μs, 	
10 Threads:   739 μs,    230 μs, 	
12 Threads:   668 μs,    197 μs,
16 Threads:   521 μs,    152 μs,
```

This is even better as with the Vector of length 100. Probably, with the extra long vectors, there occur other problems. I can imagine that the results are relatively comparable with a Vector of length 8, as @amrueda suggested above. I don't really have a strong opinion on which version is better.

Reply:
Sounds like a good compromise so far (if you comment it well enough).

Reply:
Okay perfect. I will do that :+1: 
Do you prefer to just use 8 or use @amrueda's suggestions
```
using CpuId
size_of_array = Int(ceil(cachelinesize()/sizeof(RealT)))
```
? @sloede @ranocha 
I think a hard-coded 8 would be easier to understand with appropriate comments.

Reply:
@efaulhaber just tested `cachelinesize()` on his macOS ARM. It just returns zero. So, that's not really an option.

Reply:
Huh, too bad... I wonder if there is another way to read the size of the L1 cache from Julia. On my M2 Pro, the cache line size is 128B:
```shell
% sysctl -a | grep cachelinesize
hw.cachelinesize: 128
```
This means that using `8` would lead to false sharing...

Reply:
I don't know of any processors with caches over 128B, so using 16 should be sufficient. I think I'll just hardcode it.

Reply:
I thought about that again. When using for instance `Float32` instead of `Float64` we need another stride size, right? (At least, when I use your suggestion with the division.)
I implemented a general version in [b63e787](https://github.com/trixi-framework/Trixi.jl/pull/1736/commits/b63e787a1786ecf464717b77a0d2b3dc679dd7f1).
What do you think about that @amrueda ?

Reply:
I like that solution!

Reply:
Done in [b63e787](https://github.com/trixi-framework/Trixi.jl/pull/1736/commits/b63e787a1786ecf464717b77a0d2b3dc679dd7f1)